abort-pending-builds-based-on-string-searched-on-logs.groovy

```diff
@@ -0,0 +1,38 @@
+findItems(jenkins.model.Jenkins.instance.items.findAll());
+
+def findItems(items) {
+  for (item in items) {
+    switch(item) {
+        case org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject:
+            item.getItems()?.each { itemMultibranch ->
+                searchStringInLogBuild(itemMultibranch, "String that you are looking for")
+            }
+            break;
+        case org.jenkinsci.plugins.workflow.job.WorkflowJob:
+            searchStringInLogBuild(item, "String that you are looking for")
+            break;
+
+        // you can add other kinds of Items using a case clause, or even a default value.
+
+        case com.cloudbees.hudson.plugins.folder.Folder:
+            findItems(item.getItems());
+            break;
+    }
+  }
+}
+
+def searchStringInLogBuild(item, string) {
+  item.getBuilds()?.each { build ->
+    if (build.isBuilding()) {
+      build.getLog()?.eachLine { line ->
```

Comment:
I would strongly recommend a less memory consuming approach using otherwise this script is very likely to cause serious memory performance issue son the controller:

```
BufferedReader reader = new BufferedReader(build.getLogReader());
for(String line = reader.readLine(); line != null; line = reader.readLine()) {
    logLines.add(line);
    ++lineCount;
}
try{
    BufferedReader reader = new BufferedReader(build.getLogReader());
    for (String line = br.readLine(); line != null; line = br.readLine()) {
      if (line =~ string) {
        println "Build being executed and it's going to be aborted: $build.project in build Nº$build.number"
        build.finish(hudson.model.Result.ABORTED, new java.io.IOException("Aborting build"))
        println "-------------"
      }
    }
}
catch(Exception e){
    println("Error: " + e);
} finally {
    br.close();
}
```

Reply:
Thanks for the suggestion of using a BufferedReader! Really appreciate it